Sources/TensorFlow/Operators/Basic.swift

```diff
@@ -310,6 +310,100 @@ public extension Tensor {
+    /// ```
+    /// result[p_0,          ..., p_{axis-1},
+    ///        p_{axis + 1}, ..., p_{N-1}] =
+    /// self[p_0,          ..., p_{axis-1},
+    ///      indices,
+    ///      p_{axis + 1}, ..., p_{N-1}]
+    /// ```
+    ///
+    /// For 1-D (vector) `indices`:
+    /// ```
+    /// result[p_0,          ..., p_{axis-1},
+    ///        i,
+    ///        p_{axis + 1}, ..., p_{N-1}] =
+    /// self[p_0,          ..., p_{axis-1},
+    ///      indices[i],
+    ///      p_{axis + 1}, ..., p_{N-1}]
+    /// ```
+    ///
+    /// In the general case, produces a resulting tensor where:
+    /// ```
+    /// result[p_0,             ..., p_{axis-1},
+    ///        i_{batch\_dims}, ..., i_{M-1},
+    ///        p_{axis + 1},    ..., p_{N-1}] =
+    /// self[p_0,             ..., p_{axis-1},
+    ///      indices[i_0,     ..., i_{M-1}],
+    ///      p_{axis + 1},    ..., p_{N-1}]
+    /// ```
+    /// where `N = self.rank` and `M = indices.rank`.
+    ///
+    /// The shape of the resulting tensor is:
+    /// `self.shape[..<axis] + indices.shape + self.shape[(axis + 1)...]`.
+    ///
+    /// - Note: On CPU, if an out-of-range index is found, an error is thrown. On GPU, if an
+    /// out-of-range index is found, a 0 is stored in the corresponding output values.
+    ///
+    /// - Parameters:
+    ///   - indices: Contains the indices to gather at.
+    ///   - axis: Dimension along which to gather. Negative values wrap around.
+    ///
+    /// - Precondition: `axis` must be in the range `[-rank, rank)`.
+    ///
+    /// - Returns: The gathered tensor.
+    @inlinable
+    @differentiable(wrt: self, vjp: _vjpGathering where Scalar : TensorFlowFloatingPoint)
+    func gathering(atIndices indices: Tensor<Int32>, alongAxis axis: Int = 0) -> Tensor {
+        return Raw.gatherV2(params: self, indices: indices, axis: Tensor<Int32>(Int32(axis)))
+    }
+
+    /// Gathers slices of this tensor at `indices`, while ignoring the first `batchDims` dimensions
+    /// that correspond to batch dimensions. The gather is performed along the first non-batch
+    /// dimension.
+    ///
+    /// Performs similar functionality to `gathering`, except that the resulting tensor shape is 
+    /// now:
+    /// ```
+    /// self.shape[..<batchDims] + 
+    ///   indices.shape[batchDims...] + 
+    ///   self.shape[(batchDims + indices.rank + 1)...]
+    /// ```
+    ///
+    /// - Parameters:
+    ///   - indices: Contains the indices to gather.
+    ///   - batchDims: Number of leading batch dimensions to ignore.
+    ///
+    /// - Precondition: `batchDims` must be less than `indices.rank`.
+    ///
+    /// - Returns: The gathered tensor.
+    @inlinable
+    @differentiable(wrt: self where Scalar : TensorFlowFloatingPoint)
+    func batchGathering(atIndices indices: Tensor<Int32>) -> Tensor {
+        var batchIndices = indices
+        var accumulated = Tensor<Int32>(ones: [])
+        accumulated *= shapeTensor.withoutDerivative()[1]
```

Comment:
How about using `shape` instead of `shapeTensor`? That way you won't need `withoutDerivative()`.

Reply:
We would need to eventually convert all these shapes to tensors and so that would generate a potentially significant performance penalty. I also think the same differentiability issue will come up when converting them to tensors that are used further down the line.

Reply:
I see. I also see the problem with `withoutDerivative()` being unavailable now: In conditionally differentiable functions, the tensor type does not conform to `Differentiable`.

Since the issue happens to all conditionally differentiable functions, it's necessary to introduce an global API for any type. I have two options in mind:
1. Introduce an identity function API.
```swift
@_semantics("autodiff.nonvarying")
public func withoutDerivative<T>(at x: T) -> T
```
2. Introduce a trailing closure API.
```swift
@_semantics("autodiff.nonvarying")
public func withoutDerivative<T, R>(at x: T, in f: (T) -> R) -> R
```

I'm a little torn between these.

Reply:
I see. I like the explanation on conditional differentiability. That makes sense.

In terms of usability I'd go with option 1 because I believe option 2 would often result in lots of nesting. Also, in most cases you'd like to specify that a particular tensor is not differentiable.

Reply:
Regarding the identity API, that could also be defined as a function of `Tensor`, to be "compatible" with the existing function on `Differentiable`. This could look like:
```swift
public extension Tensor {
  @inlinable
  @_semantics("autodiff.nonvarying")
  func withoutDerivative() -> Tensor { return self }
}
```